### PR TITLE
Add support for tracking map pressure for NAT map

### PR DESCRIPTION
**Pull request by derailed**
Taps into LRU maps GC cycle to surface map pressure metric based on the resulting computed map size.

Add initial map pressure metrics support for the following LRU maps:
- NAT

> Note: This is a temporary solution until LRU map counts are surfaced
in the kernel.

Related: [#20069](https://github.com/cilium/cilium/issues/20069), https://github.com/cilium/cilium/issues/21508

```release-note
Add Prometheus map pressure metrics for NAT maps
```

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

### CODE REVIEW

**Comment by ysksuzuki** [comment]
@derailed Thank you for the PR! 

Some random thoughts:

By default, the GC interval for the ct map is calculated dynamically based on maxDeleteRatio. From the GC perspective, that's OK, but it's not OK from the perspective of the user who wants to monitor the map pressure. This is not entirely accurate given that this takes time, and the map is changing while the dump occurs. I think we should use a fixed interval for the GC when the map pressure for the ct map is enabled. 
https://github.com/cilium/cilium/blob/3eaca5d3b0ac680acc2c7bf28dad9c0324b477f7/pkg/maps/ctmap/ctmap.go#L840

How do we export the map pressure of the nat map regularly? As far as I know, the flush nat map doesn't occur regularly, so we can't use the same approach as for the ct map. One possible solution is adding a go routine that exports the map pressure of the nat map regularly.

Since it is predicted that the CPU load will increase due to these changes, it seems necessary to perform them efficiently. I want to try BPF_MAP_LOOKUP_BATCH and see if we can do these efficiently.

File: pkg/maps/ctmap/ctmap.go [hunk 1]
```diff
@@ -575,11 +575,17 @@ func (f *GCFilter) doFiltering(srcIP, dstIP netip.Addr, srcPort, dstPort uint16,
 }
 
 func doGC(m *Map, filter *GCFilter) int {
+	if m.mapType.isIPv4() {
+		stats := doGC4(m, filter)
+		m.UpdatePressureMetricWithSize(int32(stats.aliveEntries))
+		return int(stats.deleted)
+	}
 	if m.mapType.isIPv6() {
```
Comment (christarazi) [inline pkg/maps/ctmap/ctmap.go]: IIUC, this code didn't exist before, as in we didn't have GC for IPv4 maps?

**Comment by christarazi** [comment]
> By default, the GC interval for the ct map is calculated dynamically based on maxDeleteRatio. From the GC perspective, that's OK, but it's not OK from the perspective of the user who wants to monitor the map pressure. This is not entirely accurate given that this takes time, and the map is changing while the dump occurs. I think we should use a fixed interval for the GC when the map pressure for the ct map is enabled.

@ysksuzuki I believe the challenge with a fixed interval is that it will put extra pressure on the CPU. Walking these maps is very expensive. While counting the entries during the GC interval is not perfect, the significant advantage is that we can now have visibility into the map entries count without any extra CPU overhead, which is a really nice property. Plus, due to the dynamic GC interval, updating the metric occurs as pressure increases on the map. So the metric is updated essentially when the user should start paying attention to the map count.

**Comment by ysksuzuki** [comment]
@christarazi 
> I believe the challenge with a fixed interval is that it will put extra pressure on the CPU. Walking these maps is very expensive on the CPU. 

I think that's why @joestringer suggests trying BPF_MAP_LOOKUP_BATCH  to perform the calculation efficiently, and I'm planning to do some PoC with it. I'm not so sure how much that could reduce the CPU cost though.

https://github.com/cilium/cilium/issues/20069
> In order to do this efficiently, this feature could make use of newer kernel features like batched map lookups (available in Linux 5.6 or later).

> the significant advantage is that we can now have visibility into the entry count without any extra CPU overhead, which is a really nice property

But we can already visualize the entry count with `cilium_datapath_conntrack_gc_entries{status="alive"}`, right? Given that we've 
already had the similar metric, I don't see the significant value in it.

Shouldn't we start by writing a design document, discussing it, and aligning our expectations for this issue?

File: pkg/maps/ctmap/ctmap.go (hunk 1 above)
Comment (derailed) [reply pkg/maps/ctmap/ctmap.go]: Tx Chris! No ipv4 gc checks was there.

**Comment by derailed** [comment]
@ysksuzuki @chris Thank you both for weighting in!

1. I did see that a size metrics are reported for both ct and nat maps but the mechanic for map_pressure is a little bit different as this metric is registerered/unregistered when over a given threshold which would be used to trigger an alert when present vs just reporting on a map size. Not to mention, consistency as it uses the very same mechanic as other map types currently use.
2. There is a correlation between ct and nat map so tho the gc cycles are different, sanitization would be performed on nat maps to ensure the entries are in sync as part of the ct gc.
3. As stated in the PR this won't accurately represent pressure but thought might be viable in the interim...
4. I think @ti-mo has already experimented with this as he has a decent prototype using BPF_MAP_LOOKUP_BATCH and got very good benchmarked results. The eternal trade off here cpu vs mem as the map need to be pre-allocated for that call to return the count. 
5. Also there is great progress per the excellent work of @aspsk to get the counts directly from the kernel which is going to be ideal but will take time to trickle down to cilium and its customers. So we need something viable in the interim...

**Comment by ysksuzuki** [comment]
@derailed Wow, some people are already working on this issue. I didn't know that. Can I see the information about those efforts somewhere? (BPF_MAP_LOOKUP_BATCH prototype and get the counts directly from the kernel)

If you need a temporary solution in the interim, I don't want to block you. Should I unassign myself from this issue for now? Please let me know if there is anything I can do to help!

**Comment by derailed** [comment]
@ysksuzuki - I'll let @ti-mo and @aspsk pipe in here with the gory details. 
As far as the issue, I think you should still own it as you are much more knowledgeable than I in this area. 
It would be great if you could review this PR and see if I missed anything for this interim solution. 

Thank you!!

File: pkg/maps/ctmap/ctmap.go [hunk 2]
```diff
@@ -575,11 +575,17 @@ func (f *GCFilter) doFiltering(srcIP, dstIP netip.Addr, srcPort, dstPort uint16,
...
-		return int(doGC6(m, filter).deleted)
-	} else if m.mapType.isIPv4() {
-		return int(doGC4(m, filter).deleted)
+		stats := doGC6(m, filter)
+		m.UpdatePressureMetricWithSize(int32(stats.aliveEntries))
+		return int(stats.deleted)
 	}
+
 	log.Fatalf("Unsupported ct map type: %s", m.mapType.String())
 	return 0
```
Comment (christarazi) [inline pkg/maps/ctmap/ctmap.go]: So it looks like the CT map GC already has some equivalent metrics tracking in `stats.finish()`, see https://github.com/cilium/cilium/blob/47724684e520154e66001bc3383eac3bccf7bde2/pkg/maps/ctmap/metrics.go#L86-L101. The metric is the one pointed out by Yusuke in https://github.com/cilium/cilium/pull/27001#issuecomment-1646525210.

So it seems we don't need these changes highlighted here.

**Review by christarazi — COMMENTED** [review]
Given the below and @ysksuzuki's comment, I can see that it doesn't quite give us much value if we just duplicate whatever we do for the existing CT GC metric and `cilium_bpf_map_pressure`. Now I understand the proposal to rather do the counting on a fixed interval and utilized batch lookups to mitigate CPU spike concerns.

**Comment by christarazi** [comment]
@ysksuzuki Does it make sense to keep this PR just for the NAT map since it doesn't have any metrics? Then in another PR, we can propose batch lookups + fixed interval counting.

**Comment by christarazi** [comment]
@derailed Thanks for updating the PR. 

I took a deep look at what this PR is doing and the NAT GC cycle. I think the changes made in `pkg/bpf/map_linux.go` are good to keep as those are the foundational pieces that we need to build off of. However, the changes in `pkg/maps/nat/nat.go` are not needed as is. The current problem with it is that the pressure metric calculation logic relies on `Flush()` being called, which upon investigation is only called from the CLI in response to a user issuing `cilium bpf ... nat flush`. Taking a closer look at how the NAT map works, we find that the datapath will insert into the map, but deletion happens during CT map GC. The fact that deletion happens during CT map GC is a bit confusing, but the code for that is [here](https://github.com/cilium/cilium/blob/551e8cd837dec24e96bb4ff95d03c5b220aee505/pkg/maps/ctmap/ctmap.go#L628). You'll [notice](https://github.com/cilium/cilium/blob/551e8cd837dec24e96bb4ff95d03c5b220aee505/pkg/maps/ctmap/ctmap.go#L662) that the delete is issued for the NAT map in that code as well. Conveniently, the logic is also [counting](https://github.com/cilium/cilium/blob/551e8cd837dec24e96bb4ff95d03c5b220aee505/pkg/maps/ctmap/ctmap.go#L666) the number of alive entries, which is exactly what we need to report as the pressure metric.

One complication though with the current implementation is that internal calls to `updatePressureMetric()` will now call `UpdatePressureMetricWithSize(len(cache))` which will overwrite the value that we would pass in from `PurgeOrphanNATEntries()`. So I think we need some mechanism to prevent this from happening. Thoughts? Once we do that, then I think this should accurately count the entries in the NAT map.

**Comment by derailed** [comment]
@christarazi Great catch! So let me see if I understood this correctly. 

The ctmap manages the content of an associated nat map during its GC cycles. 
When the purgeOrphan call is made, we end up calling DumpReliablyWithCallback which performs house keeping on the associated nat map. Tho we are tracking stats for the number of alive ingress/egress these don't seem relevant when establishing map pressure?? But since pressure metrics is now enabled on nat maps and we're maintaining a cache, deletes would trigger a [nat map pressure updates](https://github.com/cilium/cilium/blob/main/pkg/bpf/map_linux.go#L873). 
Or am I missing it?

**Comment by derailed** [comment]
@christarazi I think I understand but still unclear... From what I can tell during the orphan call we do keep a tally only for ingress/egress entries but these are only 2 of the types of entries that could be in the nat map? 
Looking at the bit flag there could be other types of entries and thus triggering the map pressure update call would not yield the correct size ie UpdatePressureWithSize(stats.IngressAlive+stats.EgressAlive). 
If this is correct,  we would need to keep another tally of alive-entries and make the call with that computed total.
Am I wrong about that?

**Comment by christarazi** [comment]
@derailed Ah, now I understand your question. It's basically, can there be more than 2 unique flags per NAT map key. From looking at the datapath code (`bpf/lib/nat.h`) for NAT and tracing the call paths, I only see `TUPLE_F_OUT` (egress) and `TUPLE_F_IN` (ingress) being set. Which makes sense given that there can only be two directions. While there's theoretically [more flags](https://github.com/cilium/cilium/blob/b3171b99441a84ad307fdf588e1c954fc7bcfd88/bpf/lib/common.h#L813-L816) that can be set, the non-directional ones are only used in the CT map keys and not the NAT map keys. It's a bit confusing because both the keys types are the same for both maps.

So in summary, 

```
UpdatePressureWithSize(stats.IngressAlive+stats.EgressAlive).
```

should be accurate.

**Comment by derailed** [comment]
@christarazi - Brilliant! Thank you for the clarification. I'll make it so.

**Comment by christarazi** [comment]
@derailed I don't see any code that addresses my previous point:

> Hence why we need to stop LRU maps from having the [automatic map pressure updates triggered](https://github.com/cilium/cilium/blob/main/pkg/bpf/map_linux.go#L873)

otherwise, the call to `UpdatePressureMetricWithSize()` inside `PurgeOrphanNATEntries()` will get overwritten.

**Comment by derailed** [comment]
@christarazi I don't think that is the case as the UpdatePressureMetricWithSize is called at the end of this process (aka after the potential deletions) and sets the metric to what's left in the map. If we don't want delete to update the metrics and given delete is implemented in the base we would need to introduce a new check to prevent the map pressure updates specifically for nats. Is this what we want?

**Comment by christarazi** [comment]
@derailed Yes, because AFAIU, LRU maps cannot be represented accurately in userspace because the datapath can trigger eviction and eviction does not trigger the userspace to update its cache. Therefore, we must only rely on walking the map, a la `PurgeOrphanNATEntries()` and a manual call to `UpdatePressureMetricWithSize()`. So I think we need logic to check the map type in https://github.com/cilium/cilium/blob/e657234cdd358572cb303236c8b81e8eeb445f7a/pkg/bpf/map_linux.go#L291 and return early if the map type is LRU.

**Comment by christarazi** [comment]
@derailed It seems that most of CI is failing presumably due to some fatal crash or something of the sort. I would suggest downloading a sysdump from one of the runs and investigate the cause. For example, this is the smoke test's sysdump: https://github.com/cilium/cilium/suites/14891492852/artifacts/848345104

**Review by christarazi — APPROVED** [review]

**Review by dylandreimerink — APPROVED** [review]

**Comment by ti-mo** [comment]
I'm not in favor of adding anything to `struct Map`, but given the lack of alternative and the change is small, it's fine for now.

**Comment by ldelossa** [comment]
all required test passing. Merging
